File: plugins/devkit/devkit-java-tests/testData/inspections/unsafeReturnStatementVisitor/NoUnsafeVisitReturnStatements.java
import com.intellij.psi.*;
import org.jetbrains.annotations.NotNull;

public class NoUnsafeVisitReturnStatements {

  // it shouldn't be reported as JavaElementVisitor is not recursive (JavaRecursiveElementWalkingVisitor/JavaRecursiveElementVisitor)
  public void methodUsingUnsupportedVisitorImplementation1(PsiElement element) {
    element.accept(new JavaElementVisitor() {
      @Override
      public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
        // do something
      }
    });
  }

  // it shouldn't be reported as it not JavaRecursiveElementWalkingVisitor/JavaRecursiveElementVisitor
  public void methodUsingUnsupportedVisitorImplementation2(PsiElement element) {
    element.accept(new MyCustomVisitor() {
      @Override
      public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
        // do something
      }
    });
  }

  // it shouldn't be reported as it implements visitClass and visitLambdaExpression
  public void methodWithImplementationContainingVisitClassAndVisitLambdaExpressionMethods1(PsiElement element) {
    element.accept(new JavaRecursiveElementVisitor() {
      @Override
      public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
        // do something
      }

      @Override
      public void visitClass(@NotNull PsiClass aClass) {
        // do something
      }

      @Override
      public void visitLambdaExpression(PsiLambdaExpression expression) {
        // do something
      }
    });
  }

  // it shouldn't be reported as it implements visitClass and visitLambdaExpression
  public void methodWithImplementationContainingVisitClassAndVisitLambdaExpressionMethods2(PsiElement element) {
    element.accept(new JavaRecursiveElementWalkingVisitor() {
      @Override
      public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
        // do something
      }

      @Override
      public void visitClass(@NotNull PsiClass aClass) {
        // do something
      }

      @Override
      public void visitLambdaExpression(PsiLambdaExpression expression) {
        // do something
      }
    });
  }

}

class MyCustomVisitor extends PsiElementVisitor {

  public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
    // do something
  }

}

// JavaElementVisitor is safe
class SafeVisitReturnStatementVisitor extends JavaElementVisitor {
  @Override
  public void visitReturnStatement(@NotNull PsiReturnStatement statement) {
    // do something
  }
}

File: plugins/devkit/devkit-kotlin-tests/testData/inspections/unsafeReturnStatementVisitor/NoUnsafeVisitReturnStatements.kt
import com.intellij.psi.*

class NoUnsafeVisitReturnStatements {
  // it shouldn't be reported as JavaElementVisitor is not recursive (JavaRecursiveElementWalkingVisitor/JavaRecursiveElementVisitor)
  fun methodUsingUnsupportedVisitorImplementation1(element: PsiElement) {
    element.accept(object : JavaElementVisitor() {
      override fun visitReturnStatement(statement: PsiReturnStatement) {
        // do something
      }
    })
  }

  // it shouldn't be reported as it not JavaRecursiveElementWalkingVisitor/JavaRecursiveElementVisitor
  fun methodUsingUnsupportedVisitorImplementation2(element: PsiElement) {
    element.accept(object : MyCustomVisitor() {
      override fun visitReturnStatement(statement: PsiReturnStatement) {
        // do something
      }
    })
  }

  // it shouldn't be reported as it implements visitClass and visitLambdaExpression
  fun methodWithImplementationContainingVisitClassAndVisitLambdaExpressionMethods1(element: PsiElement) {
    element.accept(object : JavaRecursiveElementVisitor() {
      override fun visitReturnStatement(statement: PsiReturnStatement) {
        // do something
      }

      override fun visitClass(aClass: PsiClass) {
        // do something
      }

      override fun visitLambdaExpression(expression: PsiLambdaExpression) {
        // do something
      }
    })
  }

  // it shouldn't be reported as it implements visitClass and visitLambdaExpression
  fun methodWithImplementationContainingVisitClassAndVisitLambdaExpressionMethods2(element: PsiElement) {
    element.accept(object : JavaRecursiveElementWalkingVisitor() {
      override fun visitReturnStatement(statement: PsiReturnStatement) {
        // do something
      }

      override fun visitClass(aClass: PsiClass) {
        // do something
      }

      override fun visitLambdaExpression(expression: PsiLambdaExpression) {
        // do something
      }
    })
  }
}

open class MyCustomVisitor : PsiElementVisitor() {
  open fun visitReturnStatement(statement: PsiReturnStatement) {
    // do something
  }
} // JavaElementVisitor is safe

class SafeVisitReturnStatementVisitor : JavaElementVisitor() {
  override fun visitReturnStatement(statement: PsiReturnStatement) {
    // do something
  }
}
